include/core/orderedjson.h

```diff
@@ -0,0 +1,316 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#pragma once
+
+#include <QString>
+#include <QVector>
+#include <QMap>
+#include <QPair>
+#include <QFile>
+#include <QTextStream>
+
+#include <memory>
+#include <initializer_list>
+
+#ifdef _MSC_VER
+    #if _MSC_VER <= 1800 // VS 2013
+        #ifndef noexcept
+            #define noexcept throw()
+        #endif
+
+        #ifndef snprintf
+            #define snprintf _snprintf_s
+        #endif
+    #endif
+#endif
+
+namespace poryjson {
+
+template <class T>
+class OrderedMap : public QMap<QString, T> {
+    //
+public:
+    OrderedMap() {}
+    ~OrderedMap() {}
+
+    // key value iterator
+    class iterator {
+    public:
+        iterator(const OrderedMap *map, int pos) : m_map(map), m_pos(pos) {}
+
+        bool operator != (const iterator& other) const {
+            return m_pos != other.m_pos;
+        }
+
+        bool operator == (const iterator& other) const {
+            return m_pos == other.m_pos;
+        }
+
+        iterator & operator = (const iterator &other) {
+            this->m_pos = other.m_pos;
+            this->m_map = other.m_map;
+        }
+
+        T & value () const {
+            return (*m_map)[m_map->m_keyOrder.value(m_pos)];
+        }
+
+        QPair<QString, T> operator * () const {
+            QString key = m_map->m_keyOrder.value(m_pos);
+            return QPair<QString, T>(key, m_map->value(key));
+        }
+
+        const iterator & operator ++ () {
+            m_pos++;
+            return *this;
+        }
+
+    private:
+        const OrderedMap *m_map;
+        int m_pos;
+    };
+
+    const iterator begin() const { return iterator(this, 0); }
+
+    const iterator end() const { return iterator(this, this->size()); }
+
+    iterator find(const QString &key) const {
+        int pos = 0;
+        for (auto i : *this) {
+            if (i.first == key) {
+                return iterator(this, pos);
+            }
+        }
+        return end();
+    }
+
+    bool operator < (const OrderedMap<T>& other) const {
+        return m_keyOrder.size() < other.m_keyOrder.size();
```

Comment:
Why override this method?  I would thing the existing functionality is sufficient.

Reply:
This operator isn't overloaded in the base class, but it was used in the json library so I had to add something here. I will change the functionality to mimic `std::map` (lexically on the keys afaict) though, I just wasn't sure what that was at the time.